docs/advanced_topics.rst

```diff
@@ -0,0 +1,56 @@
+This lazy allocation strategy reduces the number of total memory operations, and hence helps in cases when the memory allocation is expensive, like the pinned CPU memory or GPU memory.
+This is most visible for the operators whose output size may differ from sample to sample and from run to run. Operator with the fixed size outputs, such as crop, does not influence the overall memory consumption growth over time
+
+Operator buffer presizing
+-------------------------
+
+The purpose of this functionality is to enable the user to fine-tune the processing pipeline in situations when it is possible to forecast precisely the memory consumption during a DALI run. This results in saving the overhead of some reallocations.
+DALI uses intermediate buffers to pass data between operators in the processing graph. With DALI, the memory is never freed but just enlarged when present buffers are not sufficient to hold the data. However, in some cases, even this limited number of allocations still could affect DALI performance. Hence, if the user knows how much memory each operator buffer needs, then it is possible to provide a hint to presize buffers before the first run.
+Two parameters are available: First, the ``bytes_per_sample`` pipeline argument, which accepts one value that is used globally across all operators for all buffers.
+The second parameter is the ``bytes_per_sample_hint`` per operator argument, which accepts one value or a list of values. When one value is provided it is used for all output buffers for a given operator. When a list is provided then each buffer is presized to the corresponding size.
+
+Prefetching queue depth
+-----------------------
+
+The purpose of this functionality is to average the processing time between batches when the variation from batch to batch is high.
+DALI pipeline allows buffering one or more batches of data ahead. This becomes important when the data processing time between batches could vary a lot. Default prefetch depth is 2. The user can change this value using the ``prefetch_queue_depth`` pipeline argument. For example, if the variation is bigger then it is recommended to prefetch more ahead.
+
+C++ API
+-------
+
+.. note::
+
+  **This feature is not officially supported and may change without further notice**
+
+DALI offers two levels of C++ API. One that can be used to develop custom operators and one that allows using DALI as a library from the native applications. Both **C++ APIs are considered unstable and experimental and can be changed** without further notice. For the operator development reference please check how existing operators are implemented. For using DALI itself ``PipelineTest`` family of tests could be a good reference. 
```

Comment:
Level of C++ API is quite ambiguous. I don’t know if we can call the custom op API a C++ API. IMO this part should only focus on the  libdali C++ API calls.

Reply:
Done